### DISCUSSION TITLE
Currency on the type level

### DISCUSSION THREAD

**Discussion by jvliwanag** [dinerojs/dinero.js#407] [Ideas]
Since 2.0 is being done on typescript, wondering if it would be good to place the currency directly on the typelevel.

Currently, there's TAmount which is sort of related. But except for doing exchange rates, you'd want to be doing the currency operations (add, subtract etc) on same currencies anyway.

**Comment by sarahdayan** [comment]
Hey @jvliwanag, I'm not sure I understand what you have in mind. Could you maybe share some code examples of what you'd envision?

**Comment by jvliwanag** [comment]
Sure, so --

```typescript
const fiveUSD: Dinero<USD> = dinero({ currency: USD, amount: 5_00 });
const tenUSD: Dinero<USD> = dinero({ currency: USD, amount: 10_00 });

const eightEUR: Dinero<EUR> = dinero({ currency: EUR, amount: 8_00 });

add(fiveUSD, tenUSD); // passes type check

add(fiveUSD, eightEUR); // should not type check
```

**Reply by sarahdayan** [reply]
Interesting! Right now there are runtime assertions but I can see the value in enforcing this at build time.

I've just [POC'ed it](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQglgOwIYCcQHlIqcOB7BAHgBUBJBMAV2ABopj1qrgoBee85gPjagAoAUFChIAXBwrUaQqACNxZScAEBKNjwZNqAbgEDQkKAGEkAGwDGlUzjwoSnaj3YBvGSghIAJgVMgRnz3F4ZDRMCGxcAnslLl0AX119cGgjShR3BHMQEgBBAFs8SgRaelT0iEy-CAAPYArPAGcoBuAURABzJyhXYXcvHz9zPE8IBTKMrN1ej28EXzkkBtH6fMLiqag+2fmasAIK4AVVouB4xINoABFEcLxMSIQG3IKTumJxiqyoGrqERubWh0uj1NjMBiIXsUjpDTm4wXNBmkJiBxB9Ks81iV3kjPiBYnD+gjmuYzBAAPzQzFnPQXKDXBC3ADKyDADQAFnhgBjXqUcZVvrV6k0Wm0EJ1eCCtuCkDDKScNlKiZZypVUXystzim80Vl8dNCfMGiTTMtiMd1gIEjTknSbig8JqsTqqoK-sLAWLgQTtoMzJZrMBbKi-VYbHYzTC9aCDYM+nVxHw8GAHg1xPTbvd8I9Hdr1Xi1KweOn7TneSrdQr4fNAwApRnoAByCYLRbteGZSFZHK5EcxufLeOpSUMRjjEGLd2TWaevZOXv1Pqgxv9YeDFlDgfD5uAUcV8wIo48dQpUBIzoFv3+IqBianBFTtoZ9sz99L2IHXBbUAAbng4J4h12WwWAAMyKcwHiXMcJ0dLg+BBZcN1saRhAPMdpDiVFoLbF9s1nYpP26OFgDSBAoDAzJIM8NszzzC8hQBUVOngmRhBlPtWKXPMUOEYlSV4ZVkQAOiAhktRkTDHwzO88O3ftkU-NMaPwp08x4EFUIQQ8cHJIT4IhPsuIHOgjX4uIVESXjQRIlAyI03jEIDZDOOEcwx3EainzwHirNresGz4NR7Ks9wbLslyrIM14IqswTcR8yLhFMk0Eqsq1EriBL0qgK04j0IZHhYABVRlLjVAdCAQSg8lkcI6AAchKy56vnJdhmWRrSvqnjZEWZYAEYAAYeNEg5xAAJgwxICpaKAAFEioAJXK5FKuq2qUAahbFpaiUZCGEZxHq7bupkXqlnEIaRuqPYxMOKBJstRIKIgrN-E8N9zx+Bjr09PgxCkksVPk3EuDoeRAYdYGywUoLiNIkQhLco8IBYqz2JOcQkCRkMnJQISvE8f6hL8xtAoJmFwZJvA6zJlQKcxFQEri1VEdJgL6ZZrIJJUS18vvFhPNuATsK8taavCOCENx1ciPRgJ-vBuHEtChGkCgABqOROLy4Q8vM6aBagTx+t4IX7X0jGoSgABWQbhqM5FxCanLeZmwXxrNttLdlW37boLmUXmpbXb0Qm+BNugTd5gRw8j43xpUIA) and this looks fairly straightforward. There might be edge cases that I'm not seeing but it's worth investigating.

Feel free to open a PR if you want to 🙂

**Comment by sarahdayan** [comment]
This is now enforced in https://github.com/dinerojs/dinero.js/releases/tag/v2.0.0-alpha.17.
